### PR TITLE
Use unidirectional encryption keys in secure session

### PR DESCRIPTION
**Pull request by pan-apple**
#### Problem
As per CHIP spec, a secure session must use unidirectional encryption key. Current code is using the same key for both directions of communication.

 #### Summary of Changes
1. Generate two keys per peer connection. First key for encrypting transmitted messages, the other for decrypting received messages.
2. The keys are generated based on the role of the node during session setup. So the initiator's send key matches responder's receive key, and vice versa.

### CODE REVIEW

**Review by msandstedt — APPROVED** [review]

File: src/transport/SecureMessageCodec.cpp [hunk 1]
```diff
@@ -71,7 +71,7 @@ CHIP_ERROR Encode(NodeId localNodeId, Transport::PeerConnectionState * state, Pa
     uint16_t totalLen = msgBuf->TotalLength();
 
     MessageAuthenticationCode mac;
-    ReturnErrorOnFailure(state->GetSecureSession().Encrypt(data, totalLen, data, packetHeader, mac));
+    ReturnErrorOnFailure(state->GetSenderSecureSession().Encrypt(data, totalLen, data, packetHeader, mac));
```
Comment (bzbarsky-apple) [inline src/transport/SecureMessageCodec.cpp:74]: Maybe this is a transitory state, but what I was expecting is that there is only one secure session that has an encryption key and a decryption key and uses the appropriate one for encrypt/decrypt.

That is, deriving a secure session takes both `kSpake2pI2RSessionInfo` and `kSpake2pR2ISessionInfo` and uses them as needed to derive the two keys the session stores....

Comment (pan-apple) [reply src/transport/SecureMessageCodec.cpp:74]: The keys are derived/created only when the secure session object is created. Are you suggesting that we should move the key derivation for both keys into a single function?

Comment (bzbarsky-apple) [reply src/transport/SecureMessageCodec.cpp:74]: I understand the setup.  I'm asking whether that's the setup we want to have.  With that setup, by just writing the wrong thing in a consumer of the PeerConnectionState you can end up encrypting with the wrong key (the one for receiving), right?

Maybe the right answer is to make the secure session getters protected and expose Encrypt/Decrypt APIs on the PeerConnectionState, which then selects the right thing?

Comment (pan-apple) [reply src/transport/SecureMessageCodec.cpp]: Updated the PR to add encrypt/decrypt API in PeerConnectionState.

File: src/transport/PeerConnectionState.h [hunk 2]
```diff
@@ -77,8 +77,11 @@ class PeerConnectionState
     uint64_t GetLastActivityTimeMs() const { return mLastActivityTimeMs; }
     void SetLastActivityTimeMs(uint64_t value) { mLastActivityTimeMs = value; }
 
-    SecureSession & GetSecureSession() { return mSecureSession; }
-    const SecureSession & GetSecureSession() const { return mSecureSession; }
+    SecureSession & GetSenderSecureSession() { return mSenderSecureSession; }
+    const SecureSession & GetSenderSecureSession() const { return mSenderSecureSession; }
+
+    SecureSession & GetReceiverSecureSession() { return mReceiverSecureSession; }
+    const SecureSession & GetReceiverSecureSession() const { return mReceiverSecureSession; }
```
Comment (bzbarsky-apple) [inline src/transport/PeerConnectionState.h:84]: I guess we still need these so we can do the `DeriveSecureSession` thing?  Maybe just leave the non-const ones?

I'd still like to figure out some way to not have consumers have to think about these getters at all... maybe a followup to figure that out?

File: src/transport/PeerConnectionState.h [hunk 3]
```diff
@@ -98,7 +101,20 @@ class PeerConnectionState
         mPeerNodeId         = kUndefinedNodeId;
         mSendMessageIndex   = 0;
         mLastActivityTimeMs = 0;
-        mSecureSession.Reset();
+        mSenderSecureSession.Reset();
+        mReceiverSecureSession.Reset();
+    }
+
+    CHIP_ERROR EncryptBeforeSend(const uint8_t * input, size_t input_length, uint8_t * output, PacketHeader & header,
```
Comment (bzbarsky-apple) [inline src/transport/PeerConnectionState.h:108]: Could become const once https://github.com/project-chip/connectedhomeip/issues/4893 is fixed.  So could DecryptOnReceive.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/transport/PeerConnectionState.h (hunk 2 above)
Comment (pan-apple) [reply src/transport/PeerConnectionState.h]: Removed the const getters.

File: src/transport/PeerConnectionState.h (hunk 3 above)
Comment (pan-apple) [reply src/transport/PeerConnectionState.h:105]: The latest commit should fix #4893

**Comment by pan-apple** [comment]
Fixes #4893

**Review by andy31415 — APPROVED** [review]

**Pull request merged**